rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+  PrimitivePartitionRule
+    OnlyValidPartitionRule
+      UnionPartitionRule
+        DFPatternPartitionRule(pattern1)
+                     :
+        DFPatternPartitionRule(patternn)
+  ```
+
+### PartitionSpec
+
+A `PartitionSpec` pairs a a `PartitionRule` with one or more `Target`s.
+
+### Phase 1
+
+We build on the existing TVM support for heterogeneous devices and targets. The available `Targets` are extracted from
+the compilation configuration (eg using the existing `CompilationConfig` helper class). Each target is inspected to
+decide on how to construct a `PartitionRule`, which will guide Collage in the selection of candidate kernels to explore
+for that target. (See Appendix G for the requirements which motivated this part of the design.)
+
+- If the `Target` has a `"partition_rule"` attribute, use that directly. This would allow users to directly control
+  partitioning/fusion for the target's they care about.
+- If the `Target` has a `"compiler"` attribute (eg `"cutlass"`), and the global pattern table has an entry for that
+  attribute value, assume the `Target` denotes a pattern-based BYOC integration to explore. The `PartitionRule`
+  will import all the BYOC patterns and predicates automatically.
+- As above, but if global pattern has no matching entry, assume the `Target` denotes a predicate-based BYOC integration
+  to explore (eg `"tensorrt"`). The `PartitonRule` will look for and evaluate predicates with the
+  `"target.<compiler>"` attribute on all Relay operators.
+- Otherwise, assume the `Target` denotes a TVM-native target. The `PartitionRule` mimics `FuseOps`, but now generalized
+  to explore multiple candidates so as to leave room for possible BYOC candidates.
+
+Note that to make this approach work we need to allow for multiple `Target`s with the same `DLDeviceKind`. For the VM
+simply switching the `target` argument from dictionary to list and removing some redundant Python preprocessing code was
+all that was required to support this.
+
+The user can use `on_device` annotations to constrain sub-graphs to particular devices. When Collage is considering
+candidate partitions, it should be sure to choose a candidate `Target` which 'refines' the `Target` for every
+sub-expression discovered by the `PlanDevicesPass`. Given targets T and U we say 'T refines U' if T has a
+'"compiler"' and/or '"partition_rule"' attributes, U has no such attributes, and T and U otherwise agree on all other
+fields.
+
+### Phase 2
+
+Most of the hard work for this phase is carried by the `AllCandidates` implementations of the `PartitionRule`s. The main
+driver simply needs to index all the found `CandidatePartitions` by their minimum 'inside' `PostDfsIndex`
+for rapid retrieval during the shortest path search.
+
+### Phase 3
+
+We find it most natural to use Dijkstra to find the optimal partitioning. A `SearchState` is:
+
+- An `IndexSet` of the dataflow nodes already 'covered' by candidates on the best path to this state. This is the
```

Comment:
This point somewhat contradicts/mis-defined above as the node in the Search, where it was defined as all nodes in all paths and here it is best path -- which kind of make sense.

Is the SearchState == node in the search graph ?

Reply:
Yes, made that a bit clearer. The covered set is actually all nodes included in partitions on any path to the state, of which the best is one. Updated text.